src/cipher/gadget.rs

```diff
@@ -0,0 +1,149 @@
+use super::{PoseidonCipher, CIPHER_SIZE, MESSAGE_CAPACITY};
+use dusk_plonk::constraint_system::ecc::Point;
+use dusk_plonk::prelude::*;
+use hades252::{GadgetStrategy, Strategy};
+
+/// Return a cipher provided the PoseidonCipher encryption parameters
+pub fn poseidon_cipher_gadget(
+    composer: &mut StandardComposer,
+    shared_secret: &Point,
+    nonce: Variable,
+    message: &[Variable],
+) -> [Variable; CIPHER_SIZE] {
+    let zero = composer.add_input(BlsScalar::zero());
+
+    let ks0 = *shared_secret.x();
+    let ks1 = *shared_secret.y();
+
+    let mut cipher = [zero; CIPHER_SIZE];
+    let mut state =
+        PoseidonCipher::initial_state_circuit(composer, ks0, ks1, nonce);
+
+    GadgetStrategy::new(composer).perm(&mut state);
+
+    (0..MESSAGE_CAPACITY).for_each(|i| {
+        let x = if i < message.len() { message[i] } else { zero };
+
+        state[i + 1] = composer.add(
+            (BlsScalar::one(), state[i + 1]),
+            (BlsScalar::one(), x),
+            BlsScalar::zero(),
+            BlsScalar::zero(),
+        );
+
+        cipher[i] = state[i + 1];
+    });
+
+    GadgetStrategy::new(composer).perm(&mut state);
+    cipher[MESSAGE_CAPACITY] = state[1];
+
+    cipher
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use dusk_plonk::jubjub::{dhke, ExtendedPoint, GENERATOR};
+
+    use std::ops::Mul;
+
+    #[test]
+    fn gadget() {
```

Comment:
Why not to make the test return `Result<(), Error>` ?

I think it would be nice so you don't need to `unrwap` or `expect` and you'll get the errors printed.